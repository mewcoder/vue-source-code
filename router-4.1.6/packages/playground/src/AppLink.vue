<template>
  <a
    v-if="isExternalLink"
    v-bind="attrs"
    class="router-link"
    :class="classes"
    target="_blank"
    rel="noopener noreferrer"
    :href="to"
    :tabindex="disabled ? -1 : undefined"
    :aria-disabled="disabled"
  >
    <slot />
  </a>
  <a
    v-else
    v-bind="attrs"
    class="router-link"
    :class="classes"
    :href="href"
    :tabindex="disabled ? -1 : undefined"
    :aria-disabled="disabled"
    @click="navigate"
  >
    <slot />
  </a>
</template>

<script>
import { RouterLink, START_LOCATION, useLink, useRoute } from 'vue-router'
import { computed, defineComponent, toRefs } from 'vue'

export default defineComponent({
  props: {
    ...RouterLink.props,
    disabled: Boolean,
  },

  setup(props, { attrs }) {
    const { replace, to, disabled } = toRefs(props)
    const isExternalLink = computed(
      () => typeof to.value === 'string' && to.value.startsWith('http')
    )

    const currentRoute = useRoute()

    const { route, href, isActive, isExactActive, navigate } = useLink({
      to: computed(() => (isExternalLink.value ? START_LOCATION : to.value)),
      replace,
    })

    const classes = computed(() => ({
      // allow link to be active for unrelated routes
      'router-link-active':
        isActive.value || currentRoute.path.startsWith(route.value.path),
      'router-link-exact-active':
        isExactActive.value || currentRoute.path === route.value.path,
    }))

    return { attrs, isExternalLink, href, navigate, classes, disabled }
  },
})
</script>
