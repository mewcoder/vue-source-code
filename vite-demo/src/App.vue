<script>
import HelloWorld from "./components/HelloWorld.vue";

debugger
export default {
  name: "App",
  components: {
    HelloWorld,
  },
  data() {
    return {
      message: '小华',
    }
  },
  methods: {
    handleClick() {
      this.message = "test";
    },
  },
};
</script>

<template>
  <div id="app">
    <HelloWorld :msg="message" />
    <button @click="handleClick">handleClick</button>
  </div>
</template>
