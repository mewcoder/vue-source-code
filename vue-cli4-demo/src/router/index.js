import Vue from "vue";
import VueRouter from "vue-router";
import Home from "../views/Home.vue";

Vue.use(VueRouter);

const routes = [
  {
    path: "/",
    name: "Home",
    component: Home,
  },
  {
    path: "/about",
    name: "About",
    component: () =>
      import(/* webpackChunkName: "about" */ "../views/About.vue"),
  },
  {
    path: "/hello",
    name: "Hello",
    component: () =>
      import(/* webpackChunkName: "hello" */ "../views/Hello.vue"),
  },
];

const router = new VueRouter({
  mode: "history",
  base: process.env.BASE_URL,
  routes,
});

export default router;
