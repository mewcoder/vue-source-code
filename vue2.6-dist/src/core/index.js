import Vue from './instance/index.js';
export { default } from './instance/index.js';
import { initGlobalAPI } from './global-api/index.js';
import { isServerRendering } from './util/env.js';
import { FunctionalRenderContext } from './vdom/create-functional-component.js';

initGlobalAPI(Vue);

Object.defineProperty(Vue.prototype, '$isServer', {
  get: isServerRendering
});

Object.defineProperty(Vue.prototype, '$ssrContext', {
  get () {
    /* istanbul ignore next */
    return this.$vnode && this.$vnode.ssrContext
  }
});

// expose FunctionalRenderContext for ssr runtime helper installation
Object.defineProperty(Vue, 'FunctionalRenderContext', {
  value: FunctionalRenderContext
});

Vue.version = '2.6.14';
