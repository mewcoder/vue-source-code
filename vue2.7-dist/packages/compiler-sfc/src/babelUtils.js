import { walk } from '../../../node_modules/.pnpm/estree-walker@2.0.2/node_modules/estree-walker/dist/esm/estree-walker.js';

// https://github.com/vuejs/core/blob/main/packages/compiler-core/src/babelUtils.ts
function walkIdentifiers(root, onIdentifier, onNode) {
    const parentStack = [];
    const knownIds = Object.create(null);
    const rootExp = root.type === 'Program' &&
        root.body[0].type === 'ExpressionStatement' &&
        root.body[0].expression;
    walk(root, {
        enter(node, parent) {
            parent && parentStack.push(parent);
            if (parent &&
                parent.type.startsWith('TS') &&
                parent.type !== 'TSAsExpression' &&
                parent.type !== 'TSNonNullExpression' &&
                parent.type !== 'TSTypeAssertion') {
                return this.skip();
            }
            if (onNode)
                onNode(node);
            if (node.type === 'Identifier') {
                const isLocal = !!knownIds[node.name];
                const isRefed = isReferencedIdentifier(node, parent, parentStack);
                if ((isRefed && !isLocal)) {
                    onIdentifier(node, parent, parentStack, isRefed, isLocal);
                }
            }
            else if (node.type === 'ObjectProperty' &&
                parent.type === 'ObjectPattern') {
                node.inPattern = true;
            }
            else if (isFunctionType(node)) {
                // walk function expressions and add its arguments to known identifiers
                // so that we don't prefix them
                walkFunctionParams(node, id => markScopeIdentifier(node, id, knownIds));
            }
            else if (node.type === 'BlockStatement') {
                // #3445 record block-level local variables
                walkBlockDeclarations(node, id => markScopeIdentifier(node, id, knownIds));
            }
        },
        leave(node, parent) {
            parent && parentStack.pop();
            if (node !== rootExp && node.scopeIds) {
                for (const id of node.scopeIds) {
                    knownIds[id]--;
                    if (knownIds[id] === 0) {
                        delete knownIds[id];
                    }
                }
            }
        }
    });
}
function isReferencedIdentifier(id, parent, parentStack) {
    if (!parent) {
        return true;
    }
    // is a special keyword but parsed as identifier
    if (id.name === 'arguments') {
        return false;
    }
    if (isReferenced(id, parent)) {
        return true;
    }
    // babel's isReferenced check returns false for ids being assigned to, so we
    // need to cover those cases here
    switch (parent.type) {
        case 'AssignmentExpression':
        case 'AssignmentPattern':
            return true;
        case 'ObjectPattern':
        case 'ArrayPattern':
            return isInDestructureAssignment(parent, parentStack);
    }
    return false;
}
function isInDestructureAssignment(parent, parentStack) {
    if (parent &&
        (parent.type === 'ObjectProperty' || parent.type === 'ArrayPattern')) {
        let i = parentStack.length;
        while (i--) {
            const p = parentStack[i];
            if (p.type === 'AssignmentExpression') {
                return true;
            }
            else if (p.type !== 'ObjectProperty' && !p.type.endsWith('Pattern')) {
                break;
            }
        }
    }
    return false;
}
function walkFunctionParams(node, onIdent) {
    for (const p of node.params) {
        for (const id of extractIdentifiers(p)) {
            onIdent(id);
        }
    }
}
function walkBlockDeclarations(block, onIdent) {
    for (const stmt of block.body) {
        if (stmt.type === 'VariableDeclaration') {
            if (stmt.declare)
                continue;
            for (const decl of stmt.declarations) {
                for (const id of extractIdentifiers(decl.id)) {
                    onIdent(id);
                }
            }
        }
        else if (stmt.type === 'FunctionDeclaration' ||
            stmt.type === 'ClassDeclaration') {
            if (stmt.declare || !stmt.id)
                continue;
            onIdent(stmt.id);
        }
    }
}
function extractIdentifiers(param, nodes = []) {
    switch (param.type) {
        case 'Identifier':
            nodes.push(param);
            break;
        case 'MemberExpression':
            let object = param;
            while (object.type === 'MemberExpression') {
                object = object.object;
            }
            nodes.push(object);
            break;
        case 'ObjectPattern':
            for (const prop of param.properties) {
                if (prop.type === 'RestElement') {
                    extractIdentifiers(prop.argument, nodes);
                }
                else {
                    extractIdentifiers(prop.value, nodes);
                }
            }
            break;
        case 'ArrayPattern':
            param.elements.forEach(element => {
                if (element)
                    extractIdentifiers(element, nodes);
            });
            break;
        case 'RestElement':
            extractIdentifiers(param.argument, nodes);
            break;
        case 'AssignmentPattern':
            extractIdentifiers(param.left, nodes);
            break;
    }
    return nodes;
}
function markScopeIdentifier(node, child, knownIds) {
    const { name } = child;
    if (node.scopeIds && node.scopeIds.has(name)) {
        return;
    }
    if (name in knownIds) {
        knownIds[name]++;
    }
    else {
        knownIds[name] = 1;
    }
    (node.scopeIds || (node.scopeIds = new Set())).add(name);
}
const isFunctionType = (node) => {
    return /Function(?:Expression|Declaration)$|Method$/.test(node.type);
};
const isStaticProperty = (node) => node &&
    (node.type === 'ObjectProperty' || node.type === 'ObjectMethod') &&
    !node.computed;
/**
 * Copied from https://github.com/babel/babel/blob/main/packages/babel-types/src/validators/isReferenced.ts
 * To avoid runtime dependency on @babel/types (which includes process references)
 * This file should not change very often in babel but we may need to keep it
 * up-to-date from time to time.
 *
 * https://github.com/babel/babel/blob/main/LICENSE
 *
 */
function isReferenced(node, parent, grandparent) {
    switch (parent.type) {
        // yes: PARENT[NODE]
        // yes: NODE.child
        // no: parent.NODE
        case 'MemberExpression':
        case 'OptionalMemberExpression':
            if (parent.property === node) {
                return !!parent.computed;
            }
            return parent.object === node;
        case 'JSXMemberExpression':
            return parent.object === node;
        // no: let NODE = init;
        // yes: let id = NODE;
        case 'VariableDeclarator':
            return parent.init === node;
        // yes: () => NODE
        // no: (NODE) => {}
        case 'ArrowFunctionExpression':
            return parent.body === node;
        // no: class { #NODE; }
        // no: class { get #NODE() {} }
        // no: class { #NODE() {} }
        // no: class { fn() { return this.#NODE; } }
        case 'PrivateName':
            return false;
        // no: class { NODE() {} }
        // yes: class { [NODE]() {} }
        // no: class { foo(NODE) {} }
        case 'ClassMethod':
        case 'ClassPrivateMethod':
        case 'ObjectMethod':
            if (parent.key === node) {
                return !!parent.computed;
            }
            return false;
        // yes: { [NODE]: "" }
        // no: { NODE: "" }
        // depends: { NODE }
        // depends: { key: NODE }
        case 'ObjectProperty':
            if (parent.key === node) {
                return !!parent.computed;
            }
            // parent.value === node
            return !grandparent || grandparent.type !== 'ObjectPattern';
        // no: class { NODE = value; }
        // yes: class { [NODE] = value; }
        // yes: class { key = NODE; }
        case 'ClassProperty':
            if (parent.key === node) {
                return !!parent.computed;
            }
            return true;
        case 'ClassPrivateProperty':
            return parent.key !== node;
        // no: class NODE {}
        // yes: class Foo extends NODE {}
        case 'ClassDeclaration':
        case 'ClassExpression':
            return parent.superClass === node;
        // yes: left = NODE;
        // no: NODE = right;
        case 'AssignmentExpression':
            return parent.right === node;
        // no: [NODE = foo] = [];
        // yes: [foo = NODE] = [];
        case 'AssignmentPattern':
            return parent.right === node;
        // no: NODE: for (;;) {}
        case 'LabeledStatement':
            return false;
        // no: try {} catch (NODE) {}
        case 'CatchClause':
            return false;
        // no: function foo(...NODE) {}
        case 'RestElement':
            return false;
        case 'BreakStatement':
        case 'ContinueStatement':
            return false;
        // no: function NODE() {}
        // no: function foo(NODE) {}
        case 'FunctionDeclaration':
        case 'FunctionExpression':
            return false;
        // no: export NODE from "foo";
        // no: export * as NODE from "foo";
        case 'ExportNamespaceSpecifier':
        case 'ExportDefaultSpecifier':
            return false;
        // no: export { foo as NODE };
        // yes: export { NODE as foo };
        // no: export { NODE as foo } from "foo";
        case 'ExportSpecifier':
            // @ts-expect-error
            if (grandparent === null || grandparent === void 0 ? void 0 : grandparent.source) {
                return false;
            }
            return parent.local === node;
        // no: import NODE from "foo";
        // no: import * as NODE from "foo";
        // no: import { NODE as foo } from "foo";
        // no: import { foo as NODE } from "foo";
        // no: import NODE from "bar";
        case 'ImportDefaultSpecifier':
        case 'ImportNamespaceSpecifier':
        case 'ImportSpecifier':
            return false;
        // no: import "foo" assert { NODE: "json" }
        case 'ImportAttribute':
            return false;
        // no: <div NODE="foo" />
        case 'JSXAttribute':
            return false;
        // no: [NODE] = [];
        // no: ({ NODE }) = [];
        case 'ObjectPattern':
        case 'ArrayPattern':
            return false;
        // no: new.NODE
        // no: NODE.target
        case 'MetaProperty':
            return false;
        // yes: type X = { someProperty: NODE }
        // no: type X = { NODE: OtherType }
        case 'ObjectTypeProperty':
            return parent.key !== node;
        // yes: enum X { Foo = NODE }
        // no: enum X { NODE }
        case 'TSEnumMember':
            return parent.id !== node;
        // yes: { [NODE]: value }
        // no: { NODE: value }
        case 'TSPropertySignature':
            if (parent.key === node) {
                return !!parent.computed;
            }
            return true;
    }
    return true;
}

export { extractIdentifiers, isFunctionType, isInDestructureAssignment, isReferencedIdentifier, isStaticProperty, walkBlockDeclarations, walkFunctionParams, walkIdentifiers };
