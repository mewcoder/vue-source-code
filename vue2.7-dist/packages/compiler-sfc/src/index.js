export { parse } from './parse.js';
export { compileTemplate } from './compileTemplate.js';
export { compileStyle, compileStyleAsync } from './compileStyle.js';
export { compileScript } from './compileScript.js';
export { generateCodeFrame } from '../../../src/compiler/codeframe.js';
export { rewriteDefault } from './rewriteDefault.js';
export { parseComponent } from './parseComponent.js';
