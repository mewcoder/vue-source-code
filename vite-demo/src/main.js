import Vue from "../../vue2.6-dist/platforms/web/entry-runtime-with-compiler";
// import Vue from './vue2.6'
// import Vue from 'vue'
import App from "./App.vue";

new Vue({
  render: (h) => h(App),
}).$mount("#app");
