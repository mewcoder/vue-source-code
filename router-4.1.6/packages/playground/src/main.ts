// necessary for webpack
import { createApp } from 'vue'
import type { ComponentPublicInstance } from 'vue'
import { router, routerHistory } from './router'
import { globalState } from './store'
import App from './App.vue'

declare global {
  interface Window {
    // h: HTML5History
    h: typeof routerHistory
    r: typeof router
    vm: ComponentPublicInstance
  }
}

// for testing purposes
window.h = routerHistory
window.r = router

const app = createApp(App)
app.mixin({
  beforeRouteEnter() {
    console.log('mixin enter')
  },
})

app.provide('state', globalState)
app.use(router)

window.vm = app.mount('#app')
