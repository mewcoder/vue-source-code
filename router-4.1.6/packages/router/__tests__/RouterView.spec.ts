/**
 * @jest-environment jsdom
 */
import { RouterView } from '../src/RouterView'
import { components, RouteLocationNormalizedLoose } from './utils'
import {
  START_LOCATION_NORMALIZED,
  RouteLocationNormalized,
} from '../src/types'
import { markRaw } from 'vue'
import { createMockedRoute } from './mount'
import { mockWarn } from 'jest-mock-warn'
import { mount } from '@vue/test-utils'

// to have autocompletion
function createRoutes<T extends Record<string, RouteLocationNormalizedLoose>>(
  routes: T
): T {
  let nonReactiveRoutes: T = {} as T

  for (let key in routes) {
    nonReactiveRoutes[key] = markRaw(routes[key])
    nonReactiveRoutes[key].matched.forEach(record => {
      record.leaveGuards ??= new Set()
      record.updateGuards ??= new Set()
    })
  }

  return nonReactiveRoutes
}

const props = { default: false }

const routes = createRoutes({
  root: {
    fullPath: '/',
    name: 'home',
    path: '/',
    query: {},
    params: {},
    hash: '',
    meta: {},
    matched: [
      {
        components: { default: components.Home },
        instances: {},
        enterCallbacks: {},
        path: '/',
        props,
      },
    ],
  },
  foo: {
    fullPath: '/foo',
    name: undefined,
    path: '/foo',
    query: {},
    params: {},
    hash: '',
    meta: {},
    matched: [
      {
        components: { default: components.Foo },
        instances: {},
        enterCallbacks: {},
        path: '/foo',
        props,
      },
    ],
  },
  nested: {
    fullPath: '/a',
    name: undefined,
    path: '/a',
    query: {},
    params: {},
    hash: '',
    meta: {},
    matched: [
      {
        components: { default: components.Nested },
        instances: {},
        enterCallbacks: {},
        path: '/',
        props,
      },
      {
        components: { default: components.Foo },
        instances: {},
        enterCallbacks: {},
        path: 'a',
        props,
      },
    ],
  },
  nestedNested: {
    fullPath: '/a/b',
    name: undefined,
    path: '/a/b',
    query: {},
    params: {},
    hash: '',
    meta: {},
    matched: [
      {
        components: { default: components.Nested },
        instances: {},
        enterCallbacks: {},
        path: '/',
        props,
      },
      {
        components: { default: components.Nested },
        instances: {},
        enterCallbacks: {},
        path: 'a',
        props,
      },
      {
        components: { default: components.Foo },
        instances: {},
        enterCallbacks: {},
        path: 'b',
        props,
      },
    ],
  },
  named: {
    fullPath: '/',
    name: undefined,
    path: '/',
    query: {},
    params: {},
    hash: '',
    meta: {},
    matched: [
      {
        components: { foo: components.Foo },
        instances: {},
        enterCallbacks: {},
        path: '/',
        props,
      },
    ],
  },
  withParams: {
    fullPath: '/users/1',
    name: undefined,
    path: '/users/1',
    query: {},
    params: { id: '1' },
    hash: '',
    meta: {},
    matched: [
      {
        components: { default: components.User },

        instances: {},
        enterCallbacks: {},
        path: '/users/:id',
        props: { default: true },
      },
    ],
  },
  withIdAndOther: {
    fullPath: '/props/1',
    name: undefined,
    path: '/props/1',
    query: {},
    params: { id: '1' },
    hash: '',
    meta: {},
    matched: [
      {
        components: { default: components.WithProps },

        instances: {},
        enterCallbacks: {},
        path: '/props/:id',
        props: { default: { id: 'foo', other: 'fixed' } },
      },
    ],
  },

  withFnProps: {
    fullPath: '/props/1',
    name: undefined,
    path: '/props/1',
    query: { q: 'page' },
    params: { id: '1' },
    hash: '',
    meta: {},
    matched: [
      {
        components: { default: components.WithProps },

        instances: {},
        enterCallbacks: {},
        path: '/props/:id',
        props: {
          default: (to: RouteLocationNormalized) => ({
            id: Number(to.params.id) * 2,
            other: to.query.q,
          }),
        },
      },
    ],
  },

  passthrough: {
    fullPath: '/foo',
    name: undefined,
    path: '/foo',
    query: {},
    params: {},
    hash: '',
    meta: {},
    matched: [
      {
        components: null,
        instances: {},
        enterCallbacks: {},
        path: '/',
        props,
      },
      {
        components: { default: components.Foo },
        instances: {},
        enterCallbacks: {},
        path: 'foo',
        props,
      },
    ],
  },
})

describe('RouterView', () => {
  mockWarn()

  async function factory(
    initialRoute: RouteLocationNormalizedLoose,
    props: any = {}
  ) {
    const route = createMockedRoute(initialRoute)
    const wrapper = mount(RouterView as any, {
      props,
      global: {
        provide: route.provides,
        components: { RouterView },
      },
    })

    return { route, wrapper }
  }

  it('displays current route component', async () => {
    const { wrapper } = await factory(routes.root)
    expect(wrapper.html()).toBe(`<div>Home</div>`)
  })

  it('displays named views', async () => {
    const { wrapper } = await factory(routes.named, { name: 'foo' })
    expect(wrapper.html()).toBe(`<div>Foo</div>`)
  })

  it('displays nothing when route is unmatched', async () => {
    const { wrapper } = await factory(START_LOCATION_NORMALIZED as any)
    // NOTE: I wonder if this will stay stable in future releases
    expect('Router').not.toHaveBeenWarned()
    expect(wrapper.element.childNodes).toHaveLength(0)
  })

  it('displays nested views', async () => {
    const { wrapper } = await factory(routes.nested)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('displays deeply nested views', async () => {
    const { wrapper } = await factory(routes.nestedNested)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders when the location changes', async () => {
    const { route, wrapper } = await factory(routes.root)
    expect(wrapper.html()).toBe(`<div>Home</div>`)
    await route.set(routes.foo)
    expect(wrapper.html()).toBe(`<div>Foo</div>`)
  })

  it('does not pass params as props by default', async () => {
    let noPropsWithParams = {
      ...routes.withParams,
      matched: [
        {
          components: { default: components.User },
          instances: {},
          enterCallbacks: {},
          path: '/users/:id',
          props,
        },
      ],
    }
    const { wrapper, route } = await factory(noPropsWithParams)
    expect(wrapper.html()).toBe(`<div>User: default</div>`)
    await route.set({
      ...noPropsWithParams,
      params: { id: '4' },
    })
    expect(wrapper.html()).toBe(`<div>User: default</div>`)
  })

  it('passes params as props with props: true', async () => {
    const { wrapper, route } = await factory(routes.withParams)
    expect(wrapper.html()).toBe(`<div>User: 1</div>`)
    await route.set({
      ...routes.withParams,
      params: { id: '4' },
    })
    expect(wrapper.html()).toBe(`<div>User: 4</div>`)
  })

  it('can pass an object as props', async () => {
    const { wrapper } = await factory(routes.withIdAndOther)
    expect(wrapper.html()).toBe(`<div>id:foo;other:fixed</div>`)
  })

  it('inherit attributes', async () => {
    const { wrapper } = await factory(routes.withIdAndOther, {
      'data-test': 'true',
    })
    expect(wrapper.html()).toBe(
      `<div data-test="true">id:foo;other:fixed</div>`
    )
  })

  it('can pass a function as props', async () => {
    const { wrapper } = await factory(routes.withFnProps)
    expect(wrapper.html()).toBe(`<div>id:2;other:page</div>`)
  })

  it('pass through with empty children', async () => {
    const { wrapper } = await factory(routes.passthrough)
    expect(wrapper.html()).toBe(`<div>Foo</div>`)
  })

  describe('warnings', () => {
    it('does not warn RouterView is wrapped', () => {
      const route = createMockedRoute(routes.root)
      const wrapper = mount(
        {
          template: `<div><router-view/></div>`,
        },
        {
          props: {},
          global: {
            provide: route.provides,
            components: { RouterView },
          },
        }
      )
      expect(wrapper.html()).toMatchSnapshot()
      expect('can no longer be used directly inside').not.toHaveBeenWarned()
    })

    it('warns if KeepAlive wraps a RouterView', () => {
      const route = createMockedRoute(routes.root)
      const wrapper = mount(
        {
          template: `
        <keep-alive>
          <router-view/>
        </keep-alive>
        `,
        },
        {
          props: {},
          global: {
            provide: route.provides,
            components: { RouterView },
          },
        }
      )
      expect(wrapper.html()).toBe(`<div>Home</div>`)
      expect('can no longer be used directly inside').toHaveBeenWarned()
    })

    it('warns if KeepAlive and Transition wrap a RouterView', async () => {
      const route = createMockedRoute(routes.root)
      const wrapper = mount(
        {
          template: `
        <transition>
          <keep-alive>
            <router-view/>
          </keep-alive>
        </transition>
        `,
        },
        {
          props: {},
          global: {
            stubs: {
              transition: false,
            },
            provide: route.provides,
            components: { RouterView },
          },
        }
      )
      expect(wrapper.html()).toBe(`<div>Home</div>`)
      expect('can no longer be used directly inside').toHaveBeenWarned()
    })

    it('warns if Transition wraps a RouterView', () => {
      const route = createMockedRoute(routes.root)
      const wrapper = mount(
        {
          template: `
        <transition>
          <router-view/>
        </transition>
        `,
        },
        {
          props: {},
          global: {
            stubs: {
              transition: false,
            },
            provide: route.provides,
            components: { RouterView },
          },
        }
      )
      expect(wrapper.html()).toBe(`<div>Home</div>`)
      expect('can no longer be used directly inside').toHaveBeenWarned()
    })
  })

  describe('v-slot', () => {
    async function factory(
      initialRoute: RouteLocationNormalizedLoose,
      props: any = {}
    ) {
      const route = createMockedRoute(initialRoute)
      const wrapper = await mount(RouterView as any, {
        props,
        global: {
          provide: route.provides,
          components: { RouterView },
        },
        slots: {
          default: `
            <template #default="{ route, Component }">
              <span>{{ route.name }}</span>
              <component :is="Component"/>
            </template>
            `,
        },
      })

      return { route, wrapper }
    }

    it('passes a Component and route', async () => {
      const { wrapper } = await factory(routes.root)
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('KeepAlive', () => {
    async function factory(
      initialRoute: RouteLocationNormalizedLoose,
      props: any = {}
    ) {
      const route = createMockedRoute(initialRoute)
      const wrapper = await mount(RouterView as any, {
        props,
        global: {
          provide: route.provides,
          components: { RouterView },
        },
        slots: {
          default: `
          <template #default="{ Component }">
            <keep-alive>
              <component :is="Component"/>
            </keep-alive>
          </template>`,
        },
      })

      return { route, wrapper }
    }

    it('works', async () => {
      const { route, wrapper } = await factory(routes.root)
      expect(wrapper.html()).toMatchInlineSnapshot(`"<div>Home</div>"`)
      await route.set(routes.foo)
      expect(wrapper.html()).toMatchInlineSnapshot(`"<div>Foo</div>"`)
    })
  })

  describe('Suspense', () => {
    async function factory(
      initialRoute: RouteLocationNormalizedLoose,
      props: any = {}
    ) {
      const route = createMockedRoute(initialRoute)
      const wrapper = await mount(RouterView as any, {
        props,
        global: {
          provide: route.provides,
          components: { RouterView },
        },
        slots: {
          default: `
          <template #default="{ Component }">
            <Suspense>
              <component :is="Component"/>
            </Suspense>
          </template>`,
        },
      })

      return { route, wrapper }
    }

    it('works', async () => {
      const { route, wrapper } = await factory(routes.root)
      expect(wrapper.html()).toMatchInlineSnapshot(`"<div>Home</div>"`)
      await route.set(routes.foo)
      expect(wrapper.html()).toMatchInlineSnapshot(`"<div>Foo</div>"`)
    })
  })
})
