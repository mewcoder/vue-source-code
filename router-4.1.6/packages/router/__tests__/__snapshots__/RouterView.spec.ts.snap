// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RouterView displays deeply nested views 1`] = `
"<div>
  <h2>Nested</h2>
  <div>
    <h2>Nested</h2>
    <div>Foo</div>
  </div>
</div>"
`;

exports[`RouterView displays nested views 1`] = `
"<div>
  <h2>Nested</h2>
  <div>Foo</div>
</div>"
`;

exports[`RouterView v-slot passes a Component and route 1`] = `
"<span>home</span>
<div>Home</div>"
`;

exports[`RouterView warnings does not warn RouterView is wrapped 1`] = `
"<div>
  <div>Home</div>
</div>"
`;
